dhall-docs/src/Dhall/Docs/Core.hs

```diff
@@ -25,78 +31,112 @@ import Dhall.Docs.Embedded
 import Dhall.Docs.Html
 import Dhall.Docs.Markdown
 import Dhall.Docs.Store
-import Dhall.Parser        (Header (..), ParseError (..), exprAndHeaderFromText)
-import Dhall.Src           (Src)
-import Path                (Abs, Dir, File, Path, Rel, (</>))
-import Text.Megaparsec     (ParseErrorBundle (..))
+import Dhall.Parser
+    ( Header (..)
+    , ParseError (..)
+    , exprAndHeaderFromText
+    )
+import Dhall.Src                  (Src)
+import Path                       (Abs, Dir, File, Path, Rel, (</>))
+import Text.Megaparsec            (ParseErrorBundle (..))
 
-import qualified Control.Applicative as Applicative
+import qualified Control.Applicative        as Applicative
 import qualified Control.Monad
+import qualified Control.Monad.Trans.Writer as Writer
 import qualified Data.ByteString
-import qualified Data.List.NonEmpty  as NonEmpty
-import qualified Data.Map.Strict     as Map
+import qualified Data.Either
+import qualified Data.List.NonEmpty         as NonEmpty
+import qualified Data.Map.Strict            as Map
 import qualified Data.Maybe
+import qualified Data.Maybe                 as Maybe
 import qualified Data.Text
-import qualified Data.Text.IO        as Text.IO
+import qualified Data.Text.IO               as Text.IO
+import qualified Data.Text.Lazy             as Text.Lazy
 import qualified Dhall.Core
 import qualified Lucid
 import qualified Path
 import qualified Path.IO
+import qualified System.FilePath            as FilePath
 import qualified Text.Megaparsec
 
 -- $setup
 -- >>> :set -XQuasiQuotes
--- >>> import Path (absdir, absfile)
+-- >>> import Path (reldir)
+
+-- | Handling @DocsGenWarning@s on every function is a mess, `Writer` is our way to log
+type DocGenerator = Writer [DocsGenWarning]
+
+-- | The result of the doc-generator pure component
+data GeneratedDocs = GeneratedDocs [(Path Rel File, Text)] [DocsGenWarning]
```

Comment:
```suggestion
type DocGenerator = GeneratedDocs [(Path Rel File, Text)]

-- | The result of the doc-generator pure component
data GeneratedDocs a = GeneratedDocs a [DocsGenWarning]
```

Instead of using two separate types, you can use a single type if you add a type parameter to `GeneratedDocs` and implement `instance MonadWriter [DocsGenWarning] GeneratedDocs`